### PR TITLE
Add midnight-anchored episode builder and nocturnal evaluation

### PR DESCRIPTION
**Pull request by ShivamSingal**
## Description

This PR adds infrastructure for evaluating any forecasting model on the **nocturnal hypoglycemia prediction task** — a clinically meaningful evaluation where context ends at midnight and the model forecasts 6 hours of overnight blood glucose.

This is fundamentally different from the existing sliding-window evaluation (`iter_episodes` in `holdout_eval.py`), which measures general forecast accuracy across all times of day. Midnight-anchored evaluation measures how well a model predicts the specific overnight period where hypoglycemia risk is highest. **These two evaluation modes produce different RMSE numbers and must never be compared on the same leaderboard.**

This PR is model-agnostic and must land before the Chronos-2 model class PR (#338), so that PR can build on the evaluation infrastructure established here.

### What's added

- **`src/evaluation/episode_builders.py`** — `build_midnight_episodes()` extracts context windows ending at midnight from a single patient's data. Returns a list of episode dicts, each containing the context DataFrame, ground-truth BG for the forecast horizon, and future covariate arrays.

- **`evaluate_nocturnal_forecasting()` in `holdout_eval.py`** — Orchestrates evaluation across all holdout patients. Builds episodes per patient, stacks them into a panel DataFrame, calls `model.predict()` once, then computes per-episode and per-patient metrics.

- **5 unit tests** covering core behavior, boundary conditions, NaN handling, and covariate support.

### Design decisions and rationale

**Panel DataFrame with `episode_id` (sktime convention)**

We follow the [sktime pattern](https://github.com/sktime/sktime/blob/main/sktime/forecasting/base/_base.py) where `predict()` accepts both single-series and panel (batch) data, and returns the same type as the input. When given a panel DataFrame, `predict()` returns a panel DataFrame with `episode_id` preserved — not a plain numpy array. This means the evaluation function can unpack predictions by `episode_id` (via `groupby`) rather than relying on fragile positional indexing.

Each model is responsible for conforming to this contract internally:
- Chronos-2 renames `episode_id` → `item_id` for AutoGluon's `TimeSeriesDataFrame`
- TTM maps it to `id_columns` for its `ForecastDFDataset`
- Models that don't support batch should group by `episode_id` and loop internally — that's the model's responsibility, not the evaluation function's

**DatetimeIndex stays as the index, not a named column**

The stacked panel DataFrame keeps the DatetimeIndex from each episode's context DataFrame. We don't add a separate `timestamp` column. Models access timestamps via `df.index` and can restructure internally however they need (e.g., AutoGluon's `set_index([item_id, timestamp])` expects regular columns, so Chronos-2's `predict()` would do `df.reset_index()` to convert). This avoids imposing a specific column name that all models must know about.

**Known covariates passed separately from context**

Future covariates (IOB, COB) for the forecast horizon are passed via the `known_covariates` kwarg as a separate panel DataFrame. This is necessary because the context panel contains the data the model is allowed to see, and we cannot include future ground-truth BG values in it. The covariates panel has the same `episode_id` values for alignment.

**No forward-fill on covariates**

IOB and COB values are already computed by the Hovorka biological model on the 5-minute grid. We do not apply forward-fill or zero-fill — if NaNs exist in these columns, that's a data issue we want to know about, not silently mask. Earlier versions of this code applied `ffill().fillna(0)` which was incorrect.

**No covariate coverage filtering (for now)**

An earlier version skipped episodes where covariates had <50% coverage in the context window. We removed this because (a) we haven't analyzed real episode quality yet to know what thresholds make sense, and (b) COB is naturally more sparse than IOB, so a blanket threshold could incorrectly discard valid episodes. We'll revisit filtering after analyzing the actual data.

**Overall RMSE uses concatenated predictions**

Overall RMSE is computed by concatenating all predictions and all targets across episodes, then taking RMSE of the full vector. This is consistent with how per-patient RMSE is computed (also concatenated). An earlier version used RMS-of-per-episode-RMSEs, which is mathematically different (weights each episode equally regardless of length). Since all episodes are the same length (72 steps), the difference is negligible, but concatenated is simpler and consistent.

**`predict()` output column contract**

The evaluation function expects `predict()` to return a DataFrame with a column named after the target variable (e.g., `bg_mM`). Models that internally use different column names (e.g., Chronos-2 via AutoGluon returns `mean`) are responsible for renaming in their `predict()` implementation.

### What this PR does NOT do

- Does not modify any existing model's `predict()` method — models will be updated in their own PRs to conform to the panel DataFrame contract
- Does not wire up `evaluate_nocturnal_forecasting()` to the CLI (`--eval-mode midnight` will come in a follow-up)
- Does not add covariate quality filtering — will revisit after analyzing real episode data
- The evaluation function is not yet callable with any existing model, since no model currently returns a DataFrame with `episode_id` from `predict()`. This is deliberate — we're setting the contract here, and models conform when they're updated

### Files changed

| File | Change |
|------|--------|
| `src/evaluation/episode_builders.py` | New — `build_midnight_episodes()` |
| `src/evaluation/__init__.py` | Export `build_midnight_episodes` |
| `scripts/examples/holdout_eval.py` | Import + `evaluate_nocturnal_forecasting()` |
| `tests/evaluation/__init__.py` | New — test package |
| `tests/evaluation/test_midnight_episodes.py` | New — 5 tests |

### CODE REVIEW

File: src/evaluation/episode_builders.py [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+        window_end = anchor + forecast_length * dt
+        window_index = pd.date_range(
+            window_start, window_end, freq=freq, inclusive="left"
+        )
+
+        window_df = df.reindex(window_index)[cols_to_get]
+
+        # Skip if any BG values are missing
+        if window_df[target_col].isna().any():
+            continue
```
[the remark this answers is not in the document]
Comment (RobotPsychologist) [reply src/evaluation/episode_builders.py]: Fix will be in next push.

File: src/data/diabetes_datasets/aleppo_2017/aleppo_2017.py
[the remark this answers is not in the document]
Comment (RobotPsychologist) [reply src/data/diabetes_datasets/aleppo_2017/aleppo_2017.py]: reverted back to what is in main.

**Review by RobotPsychologist — APPROVED** [review]
Looks good,I made lots of changes, there were some issues with how the batch handling was occurring, I change the code to work in a less efficient manner but we can be more confident that we are evaluating properly. A

Also we really need to switch to git LFS sooon (on me).

**Review by RobotPsychologist — APPROVED** [review]
Added precommit changes.

File: src/evaluation/episode_builders.py [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+
+from typing import Any, Dict, List, Optional, Tuple
+
+import numpy as np
+import pandas as pd
+
+# Default sampling interval for CGM data
+SAMPLING_INTERVAL_MINUTES = 5
+
+# TODO: Switch function to be "bedtime epsidodes" instead of midnight-anchored, to better capture real-world sleep
```
[the remark this answers is not in the document]
Comment (RobotPsychologist) [reply src/evaluation/episode_builders.py]: Future work
